### ISSUE TITLE
New option to skip creation of virtual enviroment

### ISSUE THREAD

**Issue by okken** [tox-dev/tox#1181]
With all of the recent enhancements, tox almost replaces the need for using external scripts or make. However, for all of the scripts that are executed, tox creates a new venv (unless I'm missing something). For many people, tox is already running inside of a virtual env. It would be cool if that venv was allowed to be used for some tasks.

**_ENHANCEMENT issue:_**

**- [x] a clear problem statement with an example**

tox is often run within a virtual environment already.
It would be great to have an option for some environments to be able to run within the existing venv that tox is being run in already, and not create a new one.

Examples for many projects that would  benefit from this option:

* Installing developer dependencies to work on a project (but install them in the outer venv)
   * create new venv, activate, install tox, run `tox -e dev`, and start developing.
* distribution related activities
  * install wheel, build sdist and wheel, install twine, push to pypi
* run lint tools on all appropriate files
* basically, all the stuff that many people put in Makefiles

**- [x] suggested change with example**

I suggest `skipvenv=true` as an option that you can place in an environment.

**- [x] if you have want help to do a PR yourself**

I would be fine with helping with this, but I don't know where to start.

**Comment by gaborbernat** [comment]
Yeah, let's allow this.

**Comment by gaborbernat** [comment]
> Installing developer dependencies to work on a project (but install them in the outer venv)
> create new venv, activate, install tox, run tox -e dev, and start developing.

What's wrong with pipx installing tox and then using tox itself to create ``tox -e dev``? or the user can install tox user/system wide. I find it a bit misleading that the first env is created by tox itself.

> distribution related activities
> install wheel, build sdist and wheel, install twine, push to pypi

this sounds like exactly what a tox env does, create venv, install twine/pip, perform there build+publish 🤔 

> run lint tools on all appropriate files

lint tools are also provisioned oftentimes as python packages, how else would they be satisfied?

That being said I can see need for this when you just want to run core python commands. 👍

**Comment by okken** [comment]
Didn’t intend tox to create the outer env. 
What would `tox -e dev` do if not install install dependencies into the outer env?

It’d be great even if it just ran core commands. 

But I still need some place to put the list of packages that are needed for a dev env to poplulate the outer env. I think it’d be cool to use tox for that.

**Comment by gaborbernat** [comment]
My fear is what happens if the outer environment is not a virtual environment, what happens if it's not writable? Will we break someone's system Python?

**Comment by okken** [comment]
I would say throw an error if there is not an outer virtual environment, if it's possible to know.
This is old, but helps with figuring out how to tell: https://stackoverflow.com/questions/1871549/determine-if-python-is-running-inside-virtualenv

A comment below has the key:

> If you are using virtualenv (github.com/pypa/virtualenv), this answer is equally correct for Python 2 or Python 3. If you are using pyvenv (legacy.python.org/dev/peps/pep-0405), a virtualenv-equivalent built into Python 3.3+ (but not the same thing as virtualenv), then it uses sys.base_prefix instead of sys.real_prefix, and sys.base_prefix always exists; outside a pyvenv it is equal to sys.prefix. – Carl Meyer May 1 '14 at 19:21

With very limited testing, this seemed to work for me:
```python
import sys

def in_virtual_env():
    if hasattr(sys, 'real_prefix'):
        return sys.real_prefix != sys.prefix
    else:
        return sys.base_prefix != sys.prefix

if in_virtual_env():
    print('in virtual env')
else:
    print('not in virtual env')
```

would have to be tested on several version of Python, and using virtualenv as well as venv.

**Comment by gaborbernat** [comment]
But in some cases it's perfectly fine to use the system interpreter, e.g. imagine you just want to use Python instead of bash commands to achieve something. On the other hand it's hard to ensure user input it's safe or not. 🤔 This potentially can cause issues.

**Comment by okken** [comment]
Ya. of course. I think using system interpreter would be fine EXCEPT for installing stuff.

It might also be fine to blanket state that you gotta use a venv for tox to run in the skipenv mode.
Since all work is now done in an virtual env, and the skipenv mode would be saying "I already got a venv, just use that', this option might be fine for lots of cases.

But also, really all we **don't** want to happen in system python is installing stuff, correct?

**Comment by gaborbernat** [comment]
@okken check out #1193

**Comment by gaborbernat** [comment]
TBH I'm also fairly worried about the ease of exploitation, therefore given that two current active maintainers (out of three) have concerns I'll reject this as won't do for now.
